Connected Nodeboxes: Add `disconnected` boxes

Thomas--S opened the pull request:
* disconnected_top
* disconnected_bottom
* disconnected_front
* disconnected_left
* disconnected_back
* disconnected_right
* disconnected (when there is *no* neighbor)
* disconnected_sides (when there are *no* neighbors to the sides)

Useful e.g. when you want a bridge node to show a handrail when there is no bridge node next to it.

~~The `disconnected` field _could_ be used to restore the old behavior of disconnected xpanes.~~ You ~would~ need `disconnected_sides` for this. ~Working at this atm.~ _Done._

Thomas--S commented:
Update: Add backwards compatibility. The backwards compatibility is only tested roughly, further tests will follow as soon as possible (maybe tomorrow).

sfan5 commented on src/nodedef.cpp:
```diff
@@ -179,6 +200,14 @@ void NodeBox::deSerialize(std::istream &is)
+		READBOXES(unconnected_back);
+		READBOXES(unconnected_right);
+		READBOXES(unconnected);
+		READBOXES(unconnected_sides);
```
shouldn't this be in an `if` too?

sofar commented:
`unconnected` sure gives me spasm linguistically, how about we just use the english word `disconnected` instead?

Thomas--S commented:
Good idea. Thanks for the suggestion.

There is also a bug in the backwards compatibility. I will try to fix it in the afternoon.

Thomas--S commented:
UPDATE:
* renames to disconnected (Thanks, sofar!)
* Added the if-clause requested by sfan5. Thanks!
* Fixed a bug in backwards compatibility.

Should actually work fine now. It would be good if someone else could confirm that the backwards compatibility works.

Thomas--S commented:
Update: Remove forgotten debug output.

sofar commented:
Looks OK to me so far, but already needs a rebase :)

Thomas--S commented:
Thanks! I'll rebase it soon.

Thomas--S commented:
Rebased now.

Thomas--S commented:
Rebased again.

sofar commented:
ok with me. Didn't test but code is trivial enough. :+1:

Thomas--S commented:
Minor update: Fix indentation in `nodedef.cpp` lines 1244 – 1250

paramat commented:
Description is not clear as to what this does.

Thomas--S commented:
Thanks, paramat, for your feedback!

This PR is rebased now.
Additionally, I updated the description and the documentation. I hope it is better understandable now.
It would be good, if a native English speaker could check if there are any grammer/spelling mistakes in the documentation.

Thanks for your efforts!

Wuzzy2 commented:
Great, I've been wanting this. Catwalks, here we come!

paramat commented on doc/lua_api.txt:
```diff
@@ -1022,6 +1022,17 @@ A nodebox is defined as any of:
         connect_right = box OR {box1, box2, ...}
+        -- The following `disconnected_*` boxes are the opposites of the
+        -- `connect_*` ones above, i.e. when a node has no suitable neighbour
+        -- on the respective side, the according disconnected box is drawn.
```
according -> corresponding

paramat commented:
Thanks that's clear, i suspected that was the feature but just wasn't sure.

SmallJoker commented on src/nodedef.cpp:
```diff
@@ -163,6 +179,14 @@ void NodeBox::deSerialize(std::istream &is)
 		READBOXES(connect_left);
 		READBOXES(connect_back);
 		READBOXES(connect_right);
+		READBOXES(disconnected_top);
```
Either needs a NodeBox serialization version increase or a check whether the stream hasn't reached the end yet. Just reading outside the original size is dangerous.

paramat commented:
@SmallJoker is this ok now?

SmallJoker commented on src/nodedef.cpp:
```diff
@@ -107,6 +115,16 @@ void NodeBox::serialize(std::ostream &os, u16 protocol_version) const
 		WRITEBOX(connect_left);
 		WRITEBOX(connect_back);
 		WRITEBOX(connect_right);
+		if (version >= 5) {
```
Superfluous check. `version` is always `5`.
If this is fixed: :+1:

Thomas--S commented:
PR updated according to the comment above.

Thanks for your efforts!

paramat commented:
What about https://github.com/minetest/minetest/pull/5642#discussion_r157366706 ?

rubenwardy commented:
> What about #5642 (comment) ?

See the if statement in the deserialisation function

Thomas--S commented:
Thanks for the clarification! I updated the PR again according to the two comments above. Hopefully it is correct this time. :)

paramat commented:
f3b9d8707691b16d9c8b6a2e9e13db54b091dc93

paramat commented:
Is there a corresponding issue to close?

Thomas--S commented:
Thank you very much for merging!

I didn't find an corresponding issue yet, but I think this was on sofar's work and wish list on the forums.
